README
Logilab's common library
========================

What's this ?
-------------

This package contains some modules used by differents Logilab's
projects.

It is released under the GNU Public License.

There is no documentation available yet but the source code should be
clean and well documented.

Designed to ease:

* handling command line options and configuration files
* writing interactive command line tools
* manipulation files and character strings
* interfacing to OmniORB
* running unit tests
* manipulating tree structures
* generating text and HTML reports
* logging
* parsing XML processing instructions
* more...


Installation
------------

Extract the tarball, jump into the created directory and run ::

	python setup.py install

For installation options, see ::

	python setup.py install --help


Provided modules
----------------

Here is a brief description of the available modules.

Modules providing high-level features
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

* `cache`, a cache implementation with a least recently used algorithm.

* `changelog`, a tiny library to manipulate our simplified ChangeLog file format.

* `clcommands`, high-level classes to define command line programs handling
  different subcommands. It is based on `configuration` to get easy command line
  / configuration file handling.

* `cli`, a base class for interactive programs using the command line.

* `configuration`, some classes to handle unified configuration from both
  command line (using optparse) and configuration file (using ConfigParser).

* `dbf`, read Visual Fox Pro DBF files

* `proc`, interface to Linux /proc.

* `umessage`, unicode email support.

* `ureports`, micro-reports, a way to create simple reports using python objects
  without care of the final formatting. ReST and html formatters are provided.


Modules providing low-level functions and structures
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

* `compat`, provides a transparent compatibility layer between different python
  versions.

* `date`, a set of date manipulation functions.

* `daemon`, a daemon function and mix-in class to properly start an Unix daemon
  process.

* `decorators`, function decorators such as cached, timed...

* `deprecation`, decorator, metaclass & all to mark functions / classes as
  deprecated or moved

* `fileutils`, some file / file path manipulation utilities.

* `graph`, graph manipulations functions such as cycle detection, bases for dot
  file generation.

* `modutils`, python module manipulation functions.

* `shellutils`, some powerful shell like functions to replace shell scripts with
  python scripts.

* `tasksqueue`, a prioritized tasks queue implementation.

* `textutils`, some text manipulation functions (ansi colorization, line wrapping,
  rest support...).

* `tree`, base class to represent tree structure, and some others to make it
  works with the visitor implementation (see below).

* `visitor`, a generic visitor pattern implementation.


Modules extending some standard modules
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

* `debugger`,  `pdb` customization.

* `logging_ext`, extensions to `logging` module such as a colorized formatter
  and an easier initialization function.

* `optik_ext`, defines some new option types (regexp, csv, color, date, etc.)
  for `optik` / `optparse`

* `xmlrpcutils`, auth support for XML-RPC


Modules extending some external modules
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

* `corbautils`, useful functions for use with the OmniORB_ CORBA library.

* `hg`, some mercurial_ utility functions.

* `pdf_ext`, pdf and fdf file manipulations, with pdftk.

* `pyro_ext`, some Pyro_ utility functions.

* `sphinx_ext`, Sphinx_ plugin defining a `autodocstring` directive.

* `vcgutils` , utilities functions to generate file readable with Georg Sander's
  vcg tool (Visualization of Compiler Graphs).


To be deprecated modules
~~~~~~~~~~~~~~~~~~~~~~~~

Those `logilab.common` modules will much probably be deprecated in future
versions:

* `testlib`: use `unittest2`_ instead
* `pytest`: use `discover`_ instead
* `interface`: use `zope.interface` if you really want this
* `table`, `xmlutils`: is that used?
* `sphinxutils`: we won't go that way imo (i == syt)


Deprecated modules
~~~~~~~~~~~~~~~~~~

Those `logilab.common` modules are only there for backward compatibility. They
can go away at anytime.

* `optparser`: use `clcommands` instead

* `adbh`, `db`, `sqlgen`: see `logilab.database`_ instead

* `contexts`: content move to `shellutils`

* `html`: deprecated without replacement


Comments, support, bug reports
------------------------------

Project page http://www.logilab.org/project/logilab-common

Use the python-projects@lists.logilab.org mailing list. Since we do not have
publicly available bug tracker yet, bug reports should be emailed
there too.

You can subscribe to this mailing list at
http://lists.logilab.org/mailman/listinfo/python-projects

Archives are available at
http://lists.logilab.org/pipermail/python-projects/


ChangeLog
ChangeLog for logilab.common
============================

--
   * shellutils: fix tempfile issue in Execute, and deprecate it
     (CVE-2014-1839)

   * pytest: use 'env' to run the python interpreter

   * graph: ensure output is ordered on node and graph ids (#202314)


2013-16-12  --  0.60.1
    * modutils:

      * don't propagate IOError when package's __init__.py file doesn't
         exist (#174606)

      * ensure file is closed, may cause pb depending on the interpreter, eg
       pypy) (#180876)
  
      * fix support for `extend_path` based nested namespace packages ;
       Report and patch by John Johnson (#177651)

    * fix some cases of failing python3 install on windows platform / cross
     compilation (#180836)



2013-07-26  --  0.60.0
    * configuration: rename option_name method into option_attrname (#140667)

    * deprecation: new DeprecationManager class (closes #108205)

    * modutils:

        - fix typo causing name error in python3 / bad message in python2
       (#136037)
        - fix python3.3 crash in file_from_modpath due to implementation
        change of imp.find_module wrt builtin modules (#137244)

    * testlib: use assertCountEqual instead of assertSameElements/assertItemsEqual
      (deprecated), fixing crash with python 3.3 (#144526)

    * graph: use codecs.open avoid crash when writing utf-8 data under python3
      (#155138)



2013-04-16  --  0.59.1
    * graph: added pruning of the recursive search tree for detecting cycles in
      graphs (closes #2469)

    * testlib: check for generators in with_tempdir (closes #117533)

    * registry:

        - select_or_none should not silent ObjectNotFound exception
        (closes #119819)
        - remove 2 accidentally introduced tabs breaking python 3 compat
        (closes #117580)

    * fix umessages test w/ python 3 and LC_ALL=C (closes #119967, report and
      patch by Ian Delaney)



2013-01-21  --  0.59.0
    * registry:

        - introduce RegistrableObject base class, mandatory to make
          classes automatically registrable, and cleanup code
          accordingly
        - introduce objid and objname methods on Registry instead of
          classid function and inlined code plus other refactorings to allow
          arbitrary objects to be registered, provided they inherit from new
          RegistrableInstance class (closes #98742)
        - deprecate usage of leading underscore to skip object registration, using
          __abstract__ explicitly is better and notion of registered object 'name'
          is now somewhat fuzzy
        - use register_all when no registration callback defined (closes #111011)

    * logging_ext: on windows, use colorama to display colored logs, if available (closes #107436)

    * packaging: remove references to ftp at logilab

    * deprecations: really check them

    * packaging: steal spec file from fedora (closes #113099)

    * packaging force python2.6 on rhel5 (closes #113099)

    * packaging Update download and project urls (closes #113099)

    * configuration: enhance merge_options function (closes #113458)

    * decorators: fix @monkeypatch decorator contract for dark corner
      cases such as monkeypatching of a callable instance: no more
      turned into an unbound method, which was broken in python 3 and
      probably not used anywhere (actually closes #104047).



2012-11-14  --  0.58.3
    * date: fix ustrftime() impl. for python3 (closes #82161, patch by Arfrever
      Frehtes Taifersar Arahesis) and encoding detection for python2 (closes
      #109740)

    * other python3 code and test fixes (closes #104047)

    * registry: Store.setdefault shouldn't raise RegistryNotFound (closes #111010)

    * table: stop encoding to iso-8859-1, use unicode (closes #105847)

    * setup: properly install additional files during build instead of install (closes #104045)



2012-07-30  --  0.58.2
    * modutils: fixes (closes #100757 and #100935)



2012-07-17  --  0.58.1
    * modutils, testlib: be more python implementation independant (closes #99493 and #99627)



2012-04-12  --  0.58.0
    * new `registry` module containing a backport of CubicWeb selectable objects registry (closes #84654)

    * testlib: DocTestCase fix builtins pollution after doctest execution.

    * shellutil: add argument to ``ProgressBar.update`` to tune cursor progression (closes #88981)

    * deprecated: new DeprecationWrapper class (closes #88942)



2012-03-22  --  0.57.2
    * texutils: apply_units raise ValueError if string isn'nt valid (closes #88808)

    * daemon: don't call putenv directly

    * pytest: do not enable extra warning other than DeprecationWarning.

    * testlib: DocTestCase fix builtins pollution after doctest execution.

    * testlib: replace sys.exit with raise ImportError (closes: #84159)

    * fix license in README

    * add trove classifiers (tell about python 3 support for pypi)



2011-10-28  --  0.57.1
    * daemon: change $HOME after dropping privileges (closes #81297)

    * compat: method_type for py3k use instance of the class to have a
      real instance method (closes: #79268)



2011-10-12  --  0.57.0
    * only install unittest2 when python version < 2.7 (closes: #76068)

    * daemon: make pidfile world-readable (closes #75968)

    * daemon: remove unused(?) DaemonMixin class

    * update compat module for callable() and method_type()

    * decorators: fix monkeypatch py3k compat (closes #75290)

    * decorators: provide a @cachedproperty decorator



2011-09-08  --  0.56.2
    * daemon: call initgroups/setgid before setuid (closes #74173)

    * decorators: @monkeypatch should produce a method object (closes #73920)

    * modutils: allow overriding of _getobj by suppressing mangling



2011-08-05  --  0.56.1
    * clcommands: #72450 --rc-file option doesn't work



2011-06-09  --  0.56.0
    * clcommands: make registration possible by class decoration


date.py
# copyright 2003-2012 LOGILAB S.A. (Paris, FRANCE), all rights reserved.
# contact http://www.logilab.fr/ -- mailto:contact@logilab.fr
#
# This file is part of logilab-common.
#
# logilab-common is free software: you can redistribute it and/or modify it under
# the terms of the GNU Lesser General Public License as published by the Free
# Software Foundation, either version 2.1 of the License, or (at your option) any
# later version.
#
# logilab-common is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more
# details.
#
# You should have received a copy of the GNU Lesser General Public License along
# with logilab-common.  If not, see <http://www.gnu.org/licenses/>.
"""Date manipulation helper functions."""
from __future__ import division

__docformat__ = "restructuredtext en"

import math
import re
import sys
from locale import getpreferredencoding
from datetime import date, time, datetime, timedelta
from time import strptime as time_strptime
from calendar import monthrange, timegm

try:
    from mx.DateTime import RelativeDateTime, Date, DateTimeType
except ImportError:
    endOfMonth = None
    DateTimeType = datetime
else:
    endOfMonth = RelativeDateTime(months=1, day=-1)

# NOTE: should we implement a compatibility layer between date representations
#       as we have in lgc.db ?

FRENCH_FIXED_HOLIDAYS = {
    'jour_an': '%s-01-01',
    'fete_travail': '%s-05-01',
    'armistice1945': '%s-05-08',
    'fete_nat': '%s-07-14',
    'assomption': '%s-08-15',
    'toussaint': '%s-11-01',
    'armistice1918': '%s-11-11',
    'noel': '%s-12-25',
    }

FRENCH_MOBILE_HOLIDAYS = {
    'paques2004': '2004-04-12',
    'ascension2004': '2004-05-20',
    'pentecote2004': '2004-05-31',

    'paques2005': '2005-03-28',
    'ascension2005': '2005-05-05',
    'pentecote2005': '2005-05-16',

    'paques2006': '2006-04-17',
    'ascension2006': '2006-05-25',
    'pentecote2006': '2006-06-05',

    'paques2007': '2007-04-09',
    'ascension2007': '2007-05-17',
    'pentecote2007': '2007-05-28',

    'paques2008': '2008-03-24',
    'ascension2008': '2008-05-01',
    'pentecote2008': '2008-05-12',

    'paques2009': '2009-04-13',
    'ascension2009': '2009-05-21',
    'pentecote2009': '2009-06-01',

    'paques2010': '2010-04-05',
    'ascension2010': '2010-05-13',
    'pentecote2010': '2010-05-24',

    'paques2011': '2011-04-25',
    'ascension2011': '2011-06-02',
    'pentecote2011': '2011-06-13',

    'paques2012': '2012-04-09',
    'ascension2012': '2012-05-17',
    'pentecote2012': '2012-05-28',
    }

# XXX this implementation cries for multimethod dispatching

def get_step(dateobj, nbdays=1):
    # assume date is either a python datetime or a mx.DateTime object
    if isinstance(dateobj, date):
        return ONEDAY * nbdays
    return nbdays # mx.DateTime is ok with integers

def datefactory(year, month, day, sampledate):
    # assume date is either a python datetime or a mx.DateTime object
    if isinstance(sampledate, datetime):
        return datetime(year, month, day)
    if isinstance(sampledate, date):
        return date(year, month, day)
    return Date(year, month, day)

def weekday(dateobj):
    # assume date is either a python datetime or a mx.DateTime object
    if isinstance(dateobj, date):
        return dateobj.weekday()
    return dateobj.day_of_week

def str2date(datestr, sampledate):
    # NOTE: datetime.strptime is not an option until we drop py2.4 compat
    year, month, day = [int(chunk) for chunk in datestr.split('-')]
    return datefactory(year, month, day, sampledate)

def days_between(start, end):
    if isinstance(start, date):
        delta = end - start
        # datetime.timedelta.days is always an integer (floored)
        if delta.seconds:
            return delta.days + 1
        return delta.days
    else:
        return int(math.ceil((end - start).days))

def get_national_holidays(begin, end):
    """return french national days off between begin and end"""
    begin = datefactory(begin.year, begin.month, begin.day, begin)
    end = datefactory(end.year, end.month, end.day, end)
    holidays = [str2date(datestr, begin)
                for datestr in FRENCH_MOBILE_HOLIDAYS.values()]
    for year in xrange(begin.year, end.year+1):
        for datestr in FRENCH_FIXED_HOLIDAYS.values():
            date = str2date(datestr % year, begin)
            if date not in holidays:
                holidays.append(date)
    return [day for day in holidays if begin <= day < end]

def add_days_worked(start, days):
    """adds date but try to only take days worked into account"""
    step = get_step(start)
    weeks, plus = divmod(days, 5)
    end = start + ((weeks * 7) + plus) * step
    if weekday(end) >= 5: # saturday or sunday
        end += (2 * step)
    end += len([x for x in get_national_holidays(start, end + step)
                if weekday(x) < 5]) * step
    if weekday(end) >= 5: # saturday or sunday
        end += (2 * step)
    return end

def nb_open_days(start, end):
    assert start <= end
    step = get_step(start)
    days = days_between(start, end)
    weeks, plus = divmod(days, 7)
    if weekday(start) > weekday(end):
        plus -= 2
    elif weekday(end) == 6:
        plus -= 1
    open_days = weeks * 5 + plus
    nb_week_holidays = len([x for x in get_national_holidays(start, end+step)
                            if weekday(x) < 5 and x < end])
    open_days -= nb_week_holidays
    if open_days < 0:
        return 0
    return open_days

def date_range(begin, end, incday=None, incmonth=None):
    """yields each date between begin and end

    :param begin: the start date
    :param end: the end date
    :param incr: the step to use to iterate over dates. Default is
                 one day.
    :param include: None (means no exclusion) or a function taking a
                    date as parameter, and returning True if the date
                    should be included.

    When using mx datetime, you should *NOT* use incmonth argument, use instead
    oneDay, oneHour, oneMinute, oneSecond, oneWeek or endOfMonth (to enumerate
    months) as `incday` argument
    """
    assert not (incday and incmonth)
    begin = todate(begin)
    end = todate(end)
    if incmonth:
        while begin < end:
            begin = next_month(begin, incmonth)
            yield begin
    else:
        incr = get_step(begin, incday or 1)
        while begin < end:
           yield begin
           begin += incr

# makes py datetime usable #####################################################

ONEDAY = timedelta(days=1)
ONEWEEK = timedelta(days=7)

try:
    strptime = datetime.strptime
except AttributeError: # py < 2.5
    from time import strptime as time_strptime
    def strptime(value, format):
        return datetime(*time_strptime(value, format)[:6])

def strptime_time(value, format='%H:%M'):
    return time(*time_strptime(value, format)[3:6])

def todate(somedate):
    """return a date from a date (leaving unchanged) or a datetime"""
    if isinstance(somedate, datetime):
        return date(somedate.year, somedate.month, somedate.day)
    assert isinstance(somedate, (date, DateTimeType)), repr(somedate)
    return somedate

def totime(somedate):
    """return a time from a time (leaving unchanged), date or datetime"""
    # XXX mx compat
    if not isinstance(somedate, time):
        return time(somedate.hour, somedate.minute, somedate.second)
    assert isinstance(somedate, (time)), repr(somedate)
    return somedate

def todatetime(somedate):
    """return a date from a date (leaving unchanged) or a datetime"""
    # take care, datetime is a subclass of date
    if isinstance(somedate, datetime):
        return somedate
    assert isinstance(somedate, (date, DateTimeType)), repr(somedate)
    return datetime(somedate.year, somedate.month, somedate.day)

def datetime2ticks(somedate):
    return timegm(somedate.timetuple()) * 1000

def ticks2datetime(ticks):
    miliseconds, microseconds = divmod(ticks, 1000)
    try:
        return datetime.fromtimestamp(miliseconds)
    except (ValueError, OverflowError):
        epoch = datetime.fromtimestamp(0)
        nb_days, seconds = divmod(int(miliseconds), 86400)
        delta = timedelta(nb_days, seconds=seconds, microseconds=microseconds)
        try:
            return epoch + delta
        except (ValueError, OverflowError):
            raise

def days_in_month(somedate):
    return monthrange(somedate.year, somedate.month)[1]

def days_in_year(somedate):
    feb = date(somedate.year, 2, 1)
    if days_in_month(feb) == 29:
        return 366
    else:
        return 365

def previous_month(somedate, nbmonth=1):
    while nbmonth:
        somedate = first_day(somedate) - ONEDAY
        nbmonth -= 1
    return somedate

def next_month(somedate, nbmonth=1):
    while nbmonth:
        somedate = last_day(somedate) + ONEDAY
        nbmonth -= 1
    return somedate

def first_day(somedate):
    return date(somedate.year, somedate.month, 1)

def last_day(somedate):
    return date(somedate.year, somedate.month, days_in_month(somedate))

def ustrftime(somedate, fmt='%Y-%m-%d'):
    """like strftime, but returns a unicode string instead of an encoded
    string which may be problematic with localized date.

    When using Python 2, encoding is guessed by locale.getpreferredencoding().
    """
    if sys.version_info >= (3, 3):
        # datetime.date.strftime() supports dates since year 1 in Python >=3.3.
        return somedate.strftime(fmt)
    else:
        try:
            if sys.version_info < (3, 0):
                encoding = getpreferredencoding(do_setlocale=False) or 'UTF-8'
                return unicode(somedate.strftime(str(fmt)), encoding)
            else:
                return somedate.strftime(fmt)
        except ValueError, exc:
            if somedate.year >= 1900:
                raise
            # datetime is not happy with dates before 1900
            # we try to work around this, assuming a simple
            # format string
            fields = {'Y': somedate.year,
                      'm': somedate.month,
                      'd': somedate.day,
                      }
            if isinstance(somedate, datetime):
                fields.update({'H': somedate.hour,
                               'M': somedate.minute,
                               'S': somedate.second})
            fmt = re.sub('%([YmdHMS])', r'%(\1)02d', fmt)
            return unicode(fmt) % fields

def utcdatetime(dt):
    if dt.tzinfo is None:
        return dt
    return datetime(*dt.utctimetuple()[:7])

def utctime(dt):
    if dt.tzinfo is None:
        return dt
    return (dt + dt.utcoffset() + dt.dst()).replace(tzinfo=None)

def datetime_to_seconds(date):
    """return the number of seconds since the begining of the day for that date
    """
    return date.second+60*date.minute + 3600*date.hour

def timedelta_to_days(delta):
    """return the time delta as a number of seconds"""
    return delta.days + delta.seconds / (3600*24)

def timedelta_to_seconds(delta):
    """return the time delta as a fraction of days"""
    return delta.days*(3600*24) + delta.seconds
